source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -45,12 +46,13 @@ class NotReadySslSocket : public Network::TransportSocket {
 } // namespace
 
 SslSocket::SslSocket(Envoy::Ssl::ContextSharedPtr ctx, InitialState state,
-                     const Network::TransportSocketOptionsSharedPtr& transport_socket_options)
+                     const Network::TransportSocketOptionsSharedPtr& transport_socket_options,
+                     Ssl::HandshakerFactoryCb handshaker_factory_cb)
     : transport_socket_options_(transport_socket_options),
-      ctx_(std::dynamic_pointer_cast<ContextImpl>(ctx)) {
-  bssl::UniquePtr<SSL> ssl = ctx_->newSsl(transport_socket_options_.get());
-  info_ = std::make_shared<SslHandshakerImpl>(std::move(ssl), ctx_, this);
-
+      ctx_(std::dynamic_pointer_cast<ContextImpl>(ctx)),
+      info_(std::dynamic_pointer_cast<SslHandshakerImpl>(
```

Comment:
I'm probably missing something, but how is this dynamic_cast safe if someone is using a custom implementation? Don't you need a virtual method to get the info off of the handshake interface since you don't know the inheritance structure of the implementation?

Reply:
My custom implementation in the test is fortunate enough to extend SslHandshakerImpl. My thought was that anyone who wanted to write their own custom implementation of Ssl::Handshaker would also want to extend the existing SslHandshakerImpl to avoid reimplementing all the fiddly methods in ConnectionInfo. But you're right, this isn't safe for anyone who writes their own impl which doesn't extend SslHandshakerImpl.

I think the right thing to do here is (a) add the relevant methods to Ssl::Handshaker (the interface) so that it's OK for (b) `info_` to be an Ssl::Handshaker*. I'll fix this forward in a new PR.

Reply:
Cool thanks. Feel free to combine with the docs PR. Up to you.

Reply:
Addressing this in https://github.com/envoyproxy/envoy/pull/13081. Docs PR to follow.